autoload/clap/common_history.vim

```diff
@@ -0,0 +1,33 @@
+" Author: liuchengxu <xuliuchengxlc@gmail.com>
+" Description: Common utilities for command/search history.
+
+let s:save_cpo = &cpoptions
+set cpoptions&vim
+
+" Derived from fzf.vim
+function! s:get_history_list(name) abort
+  let max  = histnr(a:name)
+  let fmt  = ' %'.len(string(max)).'d '
+  let list = filter(map(range(1, max), 'histget("'. a:name .'", - v:val)'), '!empty(v:val)')
+  return list
+endfunction
+
+function! clap#common_history#source(name) abort
+  let hist_list = s:get_history_list(a:name)
+  let hist_len = len(hist_list)
+  return map(hist_list, 'printf("%4d", hist_len - v:key)."  ".v:val')
+endfunction
+
+function! clap#common_history#sink(name, selected) abort
```

Comment:
Is `type` a better name?
```suggestion
function! clap#common_history#sink(type, selected) abort
```

Reply:
My original variable name is `type`, but it located in  `hist-names` section in vim help. That's why I change it to `name`.

I am okay to change it back to `type`.

Reply:
>but it located in hist-names section in vim help.

I don't follow.

Reply:
Done!

Reply:
I search the history names or types in vim help, it calls them `hist-names`.  Try `:help hist-names` you will see.

```vim
		one of:					*hist-names*
			"cmd"	 or ":"	  command line history
			"search" or "/"   search pattern history
			"expr"	 or "="   typed expression history
			"input"  or "@"	  input line history
			"debug"  or ">"   debug command history
			empty		  the current or last used history
```